lib/ecto/schema.ex

```diff
@@ -596,7 +607,8 @@ defmodule Ecto.Schema do
   """
   defmacro has_many(name, queryable, opts \\ []) do
     quote do
-      Ecto.Schema.__has_many__(__MODULE__, unquote(name), unquote(queryable), unquote(opts))
+      queryable = Ecto.Schema.__build_assoc_module__(@association_scope, unquote(queryable))
```

Comment:
We should do as little work inside of quote as possible. So all of those functions do not actually need to use `@association_scope`. Instead, inside `__has_many__` and so on, you can run:

    Module.get_attribute(module, :association_scope) |> Ecto.Schema(queryable)

I would change all functions below to do the work inside `__has_many__`, `__belongs_to__` and so on instead of inside the quote. :)